// @ngrx/store-devtools/migrations/17_0_0-beta/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = default_1;
const ts = __importStar(require("typescript"));
const schematics_1 = require("@angular-devkit/schematics");
const schematics_core_1 = require("../..//schematics-core");
function migrate() {
    return (tree) => {
        (0, schematics_core_1.visitTSSourceFiles)(tree, (sourceFile) => {
            const devtoolsImports = sourceFile.statements
                .filter(ts.isImportDeclaration)
                .filter(({ moduleSpecifier }) => moduleSpecifier.getText(sourceFile).includes('@ngrx/store-devtools'));
            if (devtoolsImports.length === 0) {
                return;
            }
            const changes = [...findAndUpdateConfigs(sourceFile)];
            (0, schematics_core_1.commitChanges)(tree, sourceFile.fileName, changes);
        });
    };
}
function findAndUpdateConfigs(sourceFile) {
    const changes = [];
    ts.forEachChild(sourceFile, (node) => find(node, changes));
    return changes;
    function find(node, changes) {
        if (ts.isPropertyAccessExpression(node) &&
            node.name.text === 'instrument' &&
            ts.isIdentifier(node.expression) &&
            node.expression.text === 'StoreDevtoolsModule' &&
            ts.isCallExpression(node.parent)) {
            if (node.parent.arguments.length) {
                const [devtoolsConfig] = node.parent.arguments;
                if (ts.isObjectLiteralExpression(devtoolsConfig)) {
                    updateConfig(sourceFile, devtoolsConfig, (change) => changes.push(change));
                }
            }
            else {
                createDevtoolsConfig(sourceFile, node.parent, (change) => changes.push(change));
            }
        }
        if (ts.isCallExpression(node) &&
            ts.isIdentifier(node.expression) &&
            node.expression.text === 'provideStoreDevtools') {
            if (node.arguments.length) {
                const [devtoolsConfig] = node.arguments;
                if (ts.isObjectLiteralExpression(devtoolsConfig)) {
                    updateConfig(sourceFile, devtoolsConfig, (change) => changes.push(change));
                }
            }
            else {
                createDevtoolsConfig(sourceFile, node, (change) => changes.push(change));
            }
        }
        ts.forEachChild(node, (childNode) => find(childNode, changes));
    }
}
function updateConfig(sourceFile, devtoolsConfig, addChange) {
    const connectOutsideZoneProperty = devtoolsConfig.properties.find((p) => ts.isPropertyAssignment(p) &&
        ts.isIdentifier(p.name) &&
        p.name.text === 'connectOutsideZone');
    if (!connectOutsideZoneProperty) {
        addConnectInZoneProperty();
    }
    else if (ts.isPropertyAssignment(connectOutsideZoneProperty)) {
        replaceConnectOutsideZoneConfig(connectOutsideZoneProperty);
    }
    function addConnectInZoneProperty() {
        const configText = devtoolsConfig.getText(sourceFile);
        const comma = !devtoolsConfig.properties.length ||
            configText
                .substring(0, configText.length - 1)
                .trim()
                .endsWith(',')
            ? ''
            : ',';
        addChange(new schematics_core_1.InsertChange(sourceFile.fileName, devtoolsConfig.getEnd() - 1, `${comma} connectInZone: true`.trim()));
    }
    function replaceConnectOutsideZoneConfig(connectOutsideZone) {
        const currentValue = connectOutsideZone.initializer
            .getText(sourceFile)
            .trim();
        addChange((0, schematics_core_1.createReplaceChange)(sourceFile, connectOutsideZone.name, 'connectOutsideZone', 'connectInZone'));
        addChange((0, schematics_core_1.createReplaceChange)(sourceFile, connectOutsideZone.initializer, currentValue, currentValue === 'true'
            ? 'false'
            : currentValue === 'false'
                ? 'true'
                : `!${currentValue}`));
    }
}
function createDevtoolsConfig(sourceFile, callExpression, addChange) {
    addChange(new schematics_core_1.InsertChange(sourceFile.fileName, callExpression.getEnd() - 1, `{connectInZone: true}`));
}
function default_1() {
    return (0, schematics_1.chain)([migrate()]);
}
