// @ngrx/effects/migrations/9_0_0/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = default_1;
const ts = __importStar(require("typescript"));
const schematics_1 = require("@angular-devkit/schematics");
const schematics_core_1 = require("../../schematics-core");
function renameErrorHandlerConfig() {
    return (tree, ctx) => {
        (0, schematics_core_1.visitTSSourceFiles)(tree, (sourceFile) => {
            const changes = replaceEffectConfigKeys(sourceFile, 'resubscribeOnError', 'useEffectsErrorHandler');
            (0, schematics_core_1.commitChanges)(tree, sourceFile.fileName, changes);
            if (changes.length) {
                ctx.logger.info(`[@ngrx/effects] Updated Effects configuration, see the migration guide (https://ngrx.io/guide/migration/v9#effects) for more info`);
            }
        });
    };
}
function replaceEffectConfigKeys(sourceFile, oldText, newText) {
    const changes = [];
    ts.forEachChild(sourceFile, (node) => {
        visitCreateEffectFunctionCreator(node, (createEffectNode) => {
            const [effectDeclaration, configNode] = createEffectNode.arguments;
            if (configNode) {
                findAndReplaceText(configNode);
            }
        });
        visitEffectDecorator(node, (effectDecoratorNode) => {
            findAndReplaceText(effectDecoratorNode);
        });
    });
    return changes;
    function findAndReplaceText(node) {
        visitIdentifierWithText(node, oldText, (match) => {
            changes.push((0, schematics_core_1.createReplaceChange)(sourceFile, match, oldText, newText));
        });
    }
}
function visitIdentifierWithText(node, text, visitor) {
    if (ts.isIdentifier(node) && node.text === text) {
        visitor(node);
    }
    ts.forEachChild(node, (childNode) => visitIdentifierWithText(childNode, text, visitor));
}
function visitEffectDecorator(node, visitor) {
    if (ts.isDecorator(node) &&
        ts.isCallExpression(node.expression) &&
        ts.isIdentifier(node.expression.expression) &&
        node.expression.expression.text === 'Effect') {
        visitor(node);
    }
    ts.forEachChild(node, (childNode) => visitEffectDecorator(childNode, visitor));
}
function visitCreateEffectFunctionCreator(node, visitor) {
    if (ts.isCallExpression(node) &&
        ts.isIdentifier(node.expression) &&
        node.expression.text === 'createEffect') {
        visitor(node);
    }
    ts.forEachChild(node, (childNode) => visitCreateEffectFunctionCreator(childNode, visitor));
}
function default_1() {
    return (0, schematics_1.chain)([renameErrorHandlerConfig()]);
}
